Macros/Macros.js

```diff
@@ -33,54 +35,163 @@ class FurnaceMacros {
         Hooks.on('preCreateChatMessage', this.preCreateChatMessage.bind(this))
         FurnacePatching.replaceMethod(Macro, "execute", this.executeMacro)
         Macro.prototype.renderContent = this.renderMacro;
+        Macro.prototype.callScriptFunction = this.callScriptMacroFunction;
+        Object.defineProperty(Macro.prototype, "canRunAsGM", { get: this.canRunAsGM });
+    }
+    ready() {
+        game.socket.on("module.furnace", this._onSocketMessage.bind(this));
+    }
+    async _onSocketMessage(message) {
+        // To run macros as GM, first we elect a GM executor.
+        // this is to prevent running the macro more than once
+        // if there are more than one logged in GM or a single GM
+        // is logged in more than once.
+        // The election is based on each GM sending a random ID, then the user
+        // choosing one (the first it receives) and asking that specific GM session
+        // to execute the macro
+        if (message.action === "ElectGMExecutor") {
+            if (!game.user.isGM) return;
+            const electionId = randomID();
+            this._GMElectionIds.push(electionId);
+            game.socket.emit("module.furnace", {
+                action: "GMElectionID",
+                requestId: message.requestId,
+                electionId
+            });
+            // Delete the election ID in case we were not chosen after 10s, to avoid a memleak
+            setTimeout(() => {
+                this._GMElectionIds = this._GMElectionIds.filter(id => id !== electionId);
+            }, 10000);
+        } else if (message.action === "GMElectionID") {
+            const resolve = this._requestResolvers[message.requestId];
+            if (resolve) {
+                delete this._requestResolvers[message.requestId];
+                resolve(message.electionId);
+            }
+        } else if (message.action === "GMExecuteMacro") {
+            if (!game.user.isGM) return;
+            if (!this._GMElectionIds.includes(message.electionId)) return;
+            this._GMElectionIds = this._GMElectionIds.filter(id => id !== message.electionId);
+
+            const macro = game.macros.get(message.macroId);
+            const user = game.users.get(message.userId);
+            const sendResponse = (error = null, result = null) => game.socket.emit("module.furnace", {
+                action: "GMMacroResult",
+                requestId: message.requestId,
+                error
+            });
+            if (!macro)
+                return sendResponse("Cannot find macro");
+            if (!user)
+                return sendResponse("Invalid user");
+            if (macro.data.type !== "script")
+                return sendResponse("Invalid macro type");
+            if (!Macros.canUseScripts(user))
```

Comment:
I would be in favor of not doing this check - since it would allow triggers which execute a script macro to trigger without giving players execute macro permission. Its a bit like having the sticky bit require that the user have the permissions that the sticky bit confers.

Reply:
I think this might be the same suggestion as https://github.com/kakaroto/fvtt-module-furnace/pull/42#discussion_r419037842 ? 
If yes, then see my response there, if not, then please clarify whichever comment I misunderstood :)